GowerMetric.py

```diff
@@ -0,0 +1,361 @@
+import os.path
+import timeit
+from pathlib import Path
+from typing import List, Optional, Union
+
+import numpy as np
+import sklearn.utils.validation
+from scipy.spatial.distance import pdist
+from scipy.cluster.hierarchy import (
+    linkage,
+    cophenet,
+    fcluster,
+)
+from scipy.optimize import minimize
+from sklearn.metrics import (
+    silhouette_score,
+)
+from sklearn.preprocessing import OrdinalEncoder, StandardScaler
+
+from utils import DataType
+
+
+class GowerMetric:
+    def __init__(
+        self,
+        dtypes: np.array,
+        ratio_scale_normalization: str = "range",
+        weights: Optional[Union[list, str]] = None,
+        precomputed_weights_file: Optional[str] = None,
+    ):
+        assert weights is None or weights == 'precomputed' or weights == 'cpcc' or type(weights) == np.ndarray or type(weights) == list
+
+        self.dtypes = dtypes  # initialize with np.array of column data types
+        self.weights = weights
+        self.precomputed_weights_file = precomputed_weights_file
+        self.ranges_: List  # values of ranges in .ratio_scale() (iqr or traditional range)
+        self.h_: List  # h values in .ratio_scale()
+        self.n_features_in_: int
+        self.ratio_scale_normalization: str = ratio_scale_normalization
+
+        # Bit masks for certain column types
+        self.ratio_scale_idx = self.dtypes == DataType.RATIO_SCALE
+        self.cat_nom_idx = (self.dtypes == DataType.CATEGORICAL_NOMINAL) | (self.dtypes == DataType.BINARY_SYMMETRIC)
+        self.bin_asym_idx = self.dtypes == DataType.BINARY_ASYMMETRIC
+
+        # Sums of columns of given type
+        self.cat_nom_num = sum(self.cat_nom_idx)
+        self.ratio_scale_num = sum(self.ratio_scale_idx)
+        self.bin_asym_num = sum(self.bin_asym_idx)
+
+    def fit(self, X):
+        assert X.shape[1] == len(self.dtypes)
+
+        self.ranges_ = np.empty(X.shape[1])
```

Comment:
Later we only process ranges for ratio scale variables, so we can just do `np.empty(len(self.ratio_scale_idx))` here, avoid indexing later and save a bit of memory

Reply:
In fast, you do this a bit lower with `ratio_cols` variable